File: stdlib/ksign/ksignst.py
# test638 : stdlib5.ksign st

import os
import threading
import hashlib

from Cryptodome.PublicKey import RSA
from Cryptodome.Signature import pss
from Cryptodome.Hash import SHA3_512

def hashf(path, res): # file pash (abspath)
    with open(path, "rb") as f:
        temp = f.read()
    if temp == b"":
        res[0] = b"\x00" * 64
    else:
        res[0] = hashlib.sha3_512(temp).digest()

def hashd(path, res): # folder hash (abspath/)
    ps = [ path + x for x in os.listdir(path) ]
    for i in range( 0, len(ps) ):
        if os.path.isdir( ps[i] ) and ps[i][-1] != "/":
            ps[i] = ps[i] + "/"
        ps[i] = bytes(ps[i], encoding="utf-8")
    ps.sort()
    if ps == [ ]:
        res[0] = b"\x00" * 64
    else:
        temp = [b""] * len(ps)
        thr = [0] * len(ps)
        ret = [ [b""] for x in range( 0, len(ps) ) ]
        for i in range( 0, len(ps) ):
            ps[i] = str(ps[i], encoding="utf-8")
            if ps[i][-1] == "/":
                thr[i] = threading.Thread( target=hashd, args=( ps[i], ret[i] ) )
            else:
                thr[i] = threading.Thread( target=hashf, args=( ps[i], ret[i] ) )
            thr[i].start()
        for i in range( 0, len(ps) ):
            thr[i].join()
            temp[i] = ret[i][0]
        res[0] = hashlib.sha3_512( b"".join(temp) ).digest()

def infod(path): # get folder info (abspath/)
    size, file, folder = 0, 0, 1
    temp = [ path + x for x in os.listdir(path) ]
    for i in range( 0, len(temp) ):
        if os.path.isdir( temp[i] ) and temp[i][-1] != "/":
            temp[i] = temp[i] + "/"
        if temp[i][-1] == "/":
            t0, t1, t2 = infod( temp[i] )
            size, file, folder = size + t0, file + t1, folder + t2
        else:
            size, file = size + os.path.getsize( temp[i] ), file + 1
    return size, file, folder

# file/folder -> 64B khash value (general path)
def khash(path):
    ret = [b""]
    path = os.path.abspath(path).replace("\\", "/")
    if os.path.isdir(path):
        if path[-1] != "/":
            path = path + "/"
        hashd(path, ret)
    else:
        hashf(path, ret)
    return ret[0]

# get folder/file num, size info (general path) (size, file, folder)
def kinfo(path):
    path = os.path.abspath(path).replace("\\", "/")
    if os.path.isdir(path):
        if path[-1] != "/":
            path = path + "/"
        return infod(path)
    else:
        return os.path.getsize(path), 1, 0

# gen N bit public, private key 2048/4096/8192
def genkey(n):
    temp = RSA.generate(n)
    private = str(temp.export_key(format='PEM', pkcs=1), encoding="utf-8")
    public = str(temp.publickey().export_key(format='PEM'), encoding="utf-8")
    return public, private

# private S + plain nB -> enc B
def sign(private, plain):
    key = RSA.import_key( bytes(private, encoding="utf-8") )
    temp = pss.new(key).sign( SHA3_512.new(plain) )
    return temp

# public S + enc B + plain nB -> T/F (True is ok)
def verify(public, enc, plain): # !!!cannot verify sign gen by golang!!!
    key = RSA.import_key( bytes(public, encoding="utf-8") )
    v = pss.new(key)
    h = SHA3_512.new(plain)
    try:
        v.verify(h, enc)
        return True
    except:
        return False

# format : PKIX (public), PKCS1 (private), PEM, PSS

File: stdlib/ksign/ksignst.go
// test639 : stdlib5.ksign st

package ksign

// go get "golang.org/x/crypto/sha3"

import (
	"bytes"
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/x509"
	"encoding/pem"
	"errors"
	"os"
	"sort"
	"stdlib5/kio"

	"golang.org/x/crypto/sha3"
)

// check if path is folder
func isdir(path string) bool {
	temp, err := os.Stat(path)
	if err == nil {
		return temp.IsDir()
	} else {
		return false
	}
}

// file pash (abspath)
func hashf(path string, ret chan []byte) {
	var v []byte
	defer func() {
		if err := recover(); err != nil {
			v = nil
		}
		ret <- v
	}()
	f, _ := kio.Open(path, "r")
	defer f.Close()
	temp, _ := kio.Read(f, -1)
	if len(temp) == 0 {
		v = make([]byte, 64)
	} else {
		h := sha3.New512()
		h.Write(temp)
		v = h.Sum(nil)
		temp = nil
		h = nil
	}
}

// folder hash (abspath/)
func hashd(path string, ret chan []byte) {
	var v []byte
	defer func() {
		if err := recover(); err != nil {
			v = nil
		}
		ret <- v
	}()
	fs, _ := os.ReadDir(path)
	nmb := make([][]byte, len(fs))
	for i, r := range fs {
		nms := path + r.Name()
		if isdir(nms) && nms[len(nms)-1] != '/' {
			nms = nms + "/"
		}
		nmb[i] = []byte(nms)
	}
	sort.Slice(nmb, func(i, j int) bool { return bytes.Compare(nmb[i], nmb[j]) < 0 })
	if len(nmb) == 0 {
		ret <- make([]byte, 64)
	} else {
		mem := make([]chan []byte, len(nmb))
		for i, r := range nmb {
			mem[i] = make(chan []byte, 1)
			nms := string(r)
			if nms[len(nms)-1] == '/' {
				go hashd(nms, mem[i])
			} else {
				go hashf(nms, mem[i])
			}
		}
		temp := make([]byte, 0, 64*len(nmb))
		for i := 0; i < len(nmb); i++ {
			temp = append(temp, <-mem[i]...)
		}
		if len(temp) == 64*len(nmb) {
			h := sha3.New512()
			h.Write(temp)
			v = h.Sum(nil)
		} else {
			v = nil
		}
	}
}

// get folder info (abspath/)
func infod(path string, ret chan int) {
	var size, folder, file int
	defer func() {
		if err := recover(); err != nil {
			size = 0
			folder = 0
			file = 0
		}
		ret <- size
		ret <- file
		ret <- folder
	}()
	fs, _ := os.ReadDir(path)
	nms := make([]string, len(fs))
	for i, r := range fs {
		temp := path + r.Name()
		if isdir(temp) && temp[len(temp)-1] != '/' {
			temp = temp + "/"
		}
		nms[i] = temp
	}
	mem := make([]chan int, 0)
	size = 0
	file = 0
	folder = 1
	for _, r := range nms {
		if r[len(r)-1] == '/' {
			com := make(chan int, 3)
			go infod(r, com)
			mem = append(mem, com)
		} else {
			size = size + kio.Size(r)
			file = file + 1
		}
	}
	for _, r := range mem {
		size = size + <-r
		file = file + <-r
		folder = folder + <-r
	}
}

// file/folder -> 64B khash value (general path)
func Khash(path string) []byte { // runtime.GC(), debug.FreeOSMemory()
	path = kio.Abs(path)
	ret := make(chan []byte, 1)
	if path[len(path)-1] == '/' {
		go hashd(path, ret)
	} else {
		go hashf(path, ret)
	}
	return <-ret
}

// get folder/file num, size info (general path) (size, file, folder)
func Kinfo(path string) (int, int, int) {
	path = kio.Abs(path)
	if path[len(path)-1] == '/' {
		ret := make(chan int, 3)
		go infod(path, ret)
		size := <-ret
		file := <-ret
		folder := <-ret
		return size, file, folder
	} else {
		return kio.Size(path), 1, 0
	}
}

// PEM formating
func fmtpem(key []byte, keyType string) string {
	block := &pem.Block{
		Type:  keyType,
		Bytes: key,
	}

	pemKey := pem.EncodeToMemory(block)
	return string(pemKey)
}

// gen N bit public, private key 2048/4096/8192
func Genkey(n int) (string, string, error) {
	key, err := rsa.GenerateKey(rand.Reader, n)
	if err != nil {
		return "", "", err
	}
	pubASN1, err := x509.MarshalPKIXPublicKey(&key.PublicKey)
	if err != nil {
		return "", "", err
	}
	priASN1 := x509.MarshalPKCS1PrivateKey(key)
	privateKey := fmtpem(priASN1, "PRIVATE KEY")
	publicKey := fmtpem(pubASN1, "PUBLIC KEY")
	return publicKey, privateKey, nil
}

// private S + plain nB -> enc B
func Sign(private string, plain []byte) ([]byte, error) {
	block, _ := pem.Decode([]byte(private))
	key, err := x509.ParsePKCS1PrivateKey(block.Bytes)
	if err != nil {
		return nil, err
	}
	hashed := sha3.Sum512(plain)
	signature, err := rsa.SignPSS(rand.Reader, key, crypto.SHA3_512, hashed[:], &rsa.PSSOptions{SaltLength: 0, Hash: crypto.SHA3_512})
	return signature, err
}

// public S + enc B + plain nB -> T/F (True is ok)
func Verify(public string, enc []byte, plain []byte) (bool, error) {
	block, _ := pem.Decode([]byte(public))
	pubInterface, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return false, err
	}
	rsaKey, ok := pubInterface.(*rsa.PublicKey)
	if !ok {
		return false, errors.New("not an RSA public key")
	}
	hashed := sha3.Sum512([]byte(plain))
	err = rsa.VerifyPSS(rsaKey, crypto.SHA3_512, hashed[:], enc, &rsa.PSSOptions{SaltLength: 0, Hash: crypto.SHA3_512})
	if err != nil {
		return false, nil
	}
	return true, nil
}

// format : PKIX (public), PKCS1 (private), PEM, PSS
